src/scope.rs

```diff
@@ -537,11 +542,10 @@ impl Service<ServiceRequest> for ScopeService {
             true
         });
 
-        if let Some(ref app_data) = self.app_data {
-            req.add_data_container(app_data.clone());
-        }
-
-        if let Some((srv, _info)) = res {
+        if let Some(((srv, app_data), _info)) = res {
+            if let Some(ref app_data) = app_data {
+                req.add_data_container(app_data.clone());
+            }
             srv.call(req)
         } else {
             self.default.call(req)
```

Comment:
there's a regression here in that default services are losing access to it's layer's data

Reply:
Its layer's data are already set by  the parent layer before calling Service::call

Reply:
At leat this is what I think. I'll write a test case for this once I get to my dev pc.

Reply:
There are already some unit tests for such case: `resource::tests::test_data_default_service` and `scope::tests::test_override_data_default_service`.

So I can confirm that this behaves as I expect: any layer's app_data are registered to and set by the parent layer's routing service.  In other words, `req.add_data_container(app_data.clone());`, in the quoted diff, sets app_data for lower layers, not the current one.